# 1. Cleaning Data

In [346]:
import numpy as np
import pandas as pd

In [377]:
companies = pd.read_csv('companies.txt', delimiter = "\t" , encoding="ISO-8859-1")
rounds2 = pd.read_csv("rounds2.csv" , encoding="ISO-8859-1")

In [348]:
companies.head()
#companies.columns

permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [349]:
rounds2.head()
#rounds2.columns

company_permalink  \
0                  /organization/-fame   
1               /ORGANIZATION/-QOUNTER   
2               /organization/-qounter   
3  /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

In [350]:
#looks like there are upper and lower case words, so uniqueness converting everything to lower() case

#companies['permalink'] = companies['permalink'].str.lower()
#companies['name'] = companies['name'].str.lower()
companies[['permalink','name']] = companies[['permalink','name']].apply(lambda x : x.astype(str).str.lower())
companies.head()

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :qounter   
2  /organization/-the-one-of-them-inc-  (the) one of them,inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [351]:
#lower() case for company_permalink for rounds df as well 
#rounds2['company_permalink'] = rounds2['company_permalink'].str.lower()
rounds2[['company_permalink']] = rounds2[['company_permalink']].apply(lambda x : x.astype(str).str.lower())
rounds2.head()

company_permalink  \
0                  /organization/-fame   
1               /organization/-qounter   
2               /organization/-qounter   
3  /organization/-the-one-of-them-inc-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

In [352]:
#rounds2.company_permalink.unique()
len(rounds2.company_permalink.unique())

66370

In [353]:
len(companies.permalink.unique())

66368

In [354]:
#encoding and decoding the stings with most suitable values for this  dataset
# At first only permalink was taken into consideration, but after looking at the geograpically diverse dataset the names also were affected so included name as well and processed as following

companies['permalink'] = companies.permalink.str.encode('utf-8').str.decode('ascii', 'ignore')
companies['name'] = companies.name.str.encode('utf-8').str.decode('ascii', 'ignore')

In [355]:
len(companies.permalink.unique())

66368

In [356]:
#doing the same as above for rounds2 df as we need to encode and decode properly to get required fields
#rounds2['company_permalink'] = rounds2.company_permalink.str.encode('utf-8').str.decode('ascii', 'ignore')

rounds2['company_permalink'] = rounds2.company_permalink.str.encode('utf-8').str.decode('ascii', 'ignore')

In [357]:
len(rounds2.company_permalink.unique())

66368

# 2. After cleaning with respect to encoding and decoding we need to do further cleaning as following

In [358]:
#checkpoint 1 
#Sl. No 1.
print(len(rounds2.company_permalink.unique()))

#Sl. No 2.
print(len(companies.permalink.unique()))

#Sl. No 3. --> could be the 'permalink' as ans

#Sl. No 4. present in rounds but not in companies   
print(len(rounds2.loc[~rounds2['company_permalink'].isin(companies['permalink']), :]))

66368
66368
0


In [359]:
# merging the two dfs
# Sl. No 5.
master_frame = pd.merge(companies, rounds2, how="inner", left_on="permalink", right_on="company_permalink")
master_frame

permalink  \
0                                     /organization/-fame   
1                                  /organization/-qounter   
2                                  /organization/-qounter   
3                     /organization/-the-one-of-them-inc-   
4                                   /organization/0-6-com   
5                          /organization/004-technologies   
6                        /organization/01games-technology   
7                     /organization/0ndine-biomedical-inc   
8                     /organization/0ndine-biomedical-inc   
9                                    /organization/0xdata   
10                                   /organization/0xdata   
11                                   /organization/0xdata   
12                                   /organization/0xdata   
13                                        /organization/1   
14                                        /organization/1   
15                                        /organization/1   
16                              /organization/1-2-3-listo   
17                                  /organization/1-4-all   
18                         /organization/1-618-technology   
19                            /organization/1-800-dentist   
20                            /organization/1-800-doctors   
21               /organization/1-800-publicrelations-inc-   
22                             /organization/1-mainstream   
23                                  /organization/1-of-99   
24                              /organization/10-20-media   
25                              /organization/10-20-media   
26                              /organization/10-20-media   
27                              /organization/10-20-media   
28                          /organization/10-minutes-with   
29                          /organization/10-minutes-with   
...                                                   ...   
114919                              /organization/zynstra   
114920                           /organization/zyomyx-inc   
114921                           /organization/zyomyx-inc   
114922                           /organization/zyomyx-inc   
114923                           /organization/zyomyx-inc   
114924                                 /organization/zype   
114925                                 /organization/zype   
114926                               /organization/zypsee   
114927                               /organization/zypsee   
114928                     /organization/zyraz-technology   
114929                     /organization/zyraz-technology   
114930                     /organization/zyraz-technology   
114931                     /organization/zyraz-technology   
114932                                /organization/zyrra   
114933                                /organization/zyrra   
114934                                /organization/zyrra   
114935                                /organization/zyrra   
114936                               /organization/zystor   
114937                           /organization/zytoprotec   
114938                                /organization/zzish   
114939                                /organization/zzish   
114940                                /organization/zzish   
114941  /organization/zznode-science-and-technology-co...   
114942                          /organization/zzzzapp-com   
114943                          /organization/zzzzapp-com   
114944                          /organization/zzzzapp-com   
114945                          /organization/zzzzapp-com   
114946                                 /organization/eron   
114947                               /organization/asys-2   
114948  /organization/novatiff-reklam-ve-tantm-hizmetl...   

                                           name  \
0                                         #fame   
1                                      :qounter   
2                                      :qounter   
3                        (the) one of them,inc.   
4         

In [360]:
len(master_frame.index)


114949

In [361]:
# removing redundant/duplicate columns
# as both colmns used for merging are there(both unique though), hence removing one of them(as duplicate)
master_frame =  master_frame.drop(['company_permalink'], axis=1) 

In [362]:
#percentage of missing column wise values

round(100*(master_frame.isnull().sum()/len(master_frame.index)),2)

permalink                   0.00
name                        0.00
homepage_url                5.34
category_list               2.97
status                      0.00
country_code                7.55
state_code                  9.52
region                      8.84
city                        8.84
founded_at                 17.85
funding_round_permalink     0.00
funding_round_type          0.00
funding_round_code         72.91
funded_at                   0.00
raised_amount_usd          17.39
dtype: float64

Now its quite clear that the columns `funding_round_code` is not that of use since its with about 73%(72.9%) missing values, not desirable in analysis. Moreover from the business aim in place, the column like `homepage_url` can seem worthless to keep as well.

Hence dropping them as part of cleaning up.

In [363]:
# dropping columns and cleaning up the master_frame df for good ;-)
master_frame = master_frame.drop(['funding_round_code', 'homepage_url'], axis=1)
master_frame.head()

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :qounter   
2               /organization/-qounter                :qounter   
3  /organization/-the-one-of-them-inc-  (the) one of them,inc.   
4                /organization/0-6-com                 0-6.com   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2  Application Platforms|Real Time|Social Network...  operating          USA   
3                                  Apps|Games|Mobile  operating          NaN   
4                                        Curated Web  operating          CHN   

  state_code      region           city  founded_at  \
0         16      Mumbai         Mumbai         NaN   
1         DE  DE - Other  Delaware City  04-09-2014   
2         DE  DE - Other  Delaware City  04-09-2014   
3        NaN         NaN            NaN         NaN   
4         22     Beijing        Beijing  01-01-2007   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

    funded_at  raised_amount_usd  
0  05-01-2015         10000000.0  
1  14-10-2014                NaN  
2  01-03-2014           700000.0  
3  30-01-2014          3406878.0  
4  19-03-2008          2000000.0

In [364]:
#checking again
round(100*(master_frame.isnull().sum()/len(master_frame.index)),2)

#now we are left out with following columns where the null values > 5%, need to have them cleaned as well, before going forward

permalink                   0.00
name                        0.00
category_list               2.97
status                      0.00
country_code                7.55
state_code                  9.52
region                      8.84
city                        8.84
founded_at                 17.85
funding_round_permalink     0.00
funding_round_type          0.00
funded_at                   0.00
raised_amount_usd          17.39
dtype: float64

In [365]:
master_frame.raised_amount_usd.describe()

count    9.495900e+04
mean     1.042687e+07
std      1.148212e+08
min      0.000000e+00
25%      3.225000e+05
50%      1.680511e+06
75%      7.000000e+06
max      2.127194e+10
Name: raised_amount_usd, dtype: float64

Please understand that the column `raised_amount_usd` is vital to this case as because we want to analyze.

Now will have to start with `raised_amount_usd`, since its too important, lets see if we can impute values or not , or will have to drop those rows having nulls from this column to comprehend.

For imputing we  need to consider parametes like mean, standerd deviation, median, min, max and much more,
The mean as shown above isaround USD 10M USD, but the std doesn't look good. The min and max can't even consider either, as they are poles apart :( . 

With this type of wide spread stats the only operation here could be dropping the NANs off, so now let's just remove the rows having NaNs in `raised_amount_usd`, as thought earlier.


In [366]:
# removing the NANs
master_frame = master_frame[~np.isnan(master_frame['raised_amount_usd'])]
round(100*(master_frame.isnull().sum()/len(master_frame.index)), 2)


permalink                   0.00
name                        0.00
category_list               1.10
status                      0.00
country_code                6.16
state_code                  8.01
region                      7.42
city                        7.42
founded_at                 16.81
funding_round_permalink     0.00
funding_round_type          0.00
funded_at                   0.00
raised_amount_usd           0.00
dtype: float64

In [367]:
master_frame['raised_amount_usd'] = master_frame['raised_amount_usd']/1000000
master_frame.raised_amount_usd.describe()
#had to convert it to million of $ range

count    94959.000000
mean        10.426869
std        114.821248
min          0.000000
25%          0.322500
50%          1.680511
75%          7.000000
max      21271.935000
Name: raised_amount_usd, dtype: float64

In [376]:
#checkpoint 2
master_frame.funding_round_type.describe()

#as its datatype is object, finding avegarge is as below 

count       94959
unique         14
top       venture
freq        50228
Name: funding_round_type, dtype: object

In [368]:
#Findding all the avgerage values over funding_round_type

master_frame.groupby(["funding_round_type"]).mean() 

raised_amount_usd
funding_round_type                      
angel                           0.958694
convertible_note                1.453439
debt_financing                 17.043526
equity_crowdfunding             0.538368
grant                           4.300576
non_equity_assistance           0.411203
post_ipo_debt                 168.704572
post_ipo_equity                82.182494
private_equity                 73.308593
product_crowdfunding            1.363131
secondary_market               79.649630
seed                            0.719818
undisclosed                    19.242370
venture                        11.748949

Referred below links to find the avg of an object in pandas DF to get an idea how the avg works on objects and after R&D found it to be very simple as above:)

https://datascience.stackexchange.com/questions/19243/calculating-mean-of-data-frame-inside-a-series-object

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.mean.html

There are more links that were reffered, not going to mentioned them all ;-)

### And looks like "venture" is suitable for Spark

In [369]:
#country_code = master_frame['country_code'].astype('category')

# displaying frequencies of each category
master_frame.country_code.value_counts()

USA    62049
GBR     5019
CAN     2616
CHN     1927
IND     1649
FRA     1451
ISR     1364
ESP     1074
DEU     1042
AUS      649
RUS      588
IRL      563
SWE      560
SGP      546
NLD      532
JPN      485
ITA      483
BRA      483
CHE      437
CHL      432
KOR      432
FIN      382
DNK      314
ARG      297
BEL      293
HKG      250
TUR      196
NOR      191
BGR      190
MEX      189
       ...  
NIC        2
MAF        2
ZMB        2
KHM        2
BLZ        2
CIV        2
ZWE        2
BAH        2
DOM        2
PRY        1
TGO        1
DZA        1
MKD        1
PSE        1
SEN        1
GGY        1
LAO        1
UZB        1
MNE        1
SOM        1
GRD        1
KNA        1
SYC        1
QAT        1
OMN        1
BLM        1
HND        1
DMA        1
IRN        1
BRB        1
Name: country_code, Length: 134, dtype: int64

USA tops here as above values shows, country category wise`

In [370]:
master_frame.info()
print('\nRetained Rows Count, have a look:(in %age)')
round(100*(len(master_frame.index)/len(rounds2.index)),2)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 94959 entries, 0 to 114948
Data columns (total 13 columns):
permalink                  94959 non-null object
name                       94959 non-null object
category_list              93915 non-null object
status                     94959 non-null object
country_code               89108 non-null object
state_code                 87355 non-null object
region                     87911 non-null object
city                       87914 non-null object
founded_at                 78993 non-null object
funding_round_permalink    94959 non-null object
funding_round_type         94959 non-null object
funded_at                  94959 non-null object
raised_amount_usd          94959 non-null float64
dtypes: float64(1), object(12)
memory usage: 10.1+ MB

Retained Rows Count, have a look:(in %age)


82.61

After looking at the results above, it looks like clean dataframe now, we will write it to a csv for safer execution of case study further,

lets do it:)

In [373]:
master_frame.to_csv("master_frame_cleaned.csv", sep=',', index=False)

This is upto the first part of the Case Study , Next is there in Part 2 files, had to do it, as its messy now:)
Excel updated as well.

please refer the other part in other .ipynb file

Thanks!

In [374]:
master_frame

permalink  \
0                                     /organization/-fame   
2                                  /organization/-qounter   
3                     /organization/-the-one-of-them-inc-   
4                                   /organization/0-6-com   
6                        /organization/01games-technology   
7                     /organization/0ndine-biomedical-inc   
8                     /organization/0ndine-biomedical-inc   
9                                    /organization/0xdata   
10                                   /organization/0xdata   
11                                   /organization/0xdata   
12                                   /organization/0xdata   
13                                        /organization/1   
15                                        /organization/1   
16                              /organization/1-2-3-listo   
20                            /organization/1-800-doctors   
21               /organization/1-800-publicrelations-inc-   
22                             /organization/1-mainstream   
23                                  /organization/1-of-99   
24                              /organization/10-20-media   
25                              /organization/10-20-media   
26                              /organization/10-20-media   
27                              /organization/10-20-media   
28                          /organization/10-minutes-with   
29                          /organization/10-minutes-with   
30                               /organization/1000-corks   
31                             /organization/1000-markets   
32                                  /organization/1000chi   
34                             /organization/1000memories   
35                             /organization/1000memories   
36                          /organization/1000museums-com   
...                                                   ...   
114917                              /organization/zynstra   
114918                              /organization/zynstra   
114919                              /organization/zynstra   
114920                           /organization/zyomyx-inc   
114921                           /organization/zyomyx-inc   
114922                           /organization/zyomyx-inc   
114923                           /organization/zyomyx-inc   
114924                                 /organization/zype   
114925                                 /organization/zype   
114926                               /organization/zypsee   
114928                     /organization/zyraz-technology   
114929                     /organization/zyraz-technology   
114930                     /organization/zyraz-technology   
114931                     /organization/zyraz-technology   
114932                                /organization/zyrra   
114933                                /organization/zyrra   
114934                                /organization/zyrra   
114935                                /organization/zyrra   
114936                               /organization/zystor   
114937                           /organization/zytoprotec   
114938                                /organization/zzish   
114939                                /organization/zzish   
114940                                /organization/zzish   
114941  /organization/zznode-science-and-technology-co...   
114942                          /organization/zzzzapp-com   
114943                          /organization/zzzzapp-com   
114944                          /organization/zzzzapp-com   
114945                          /organization/zzzzapp-com   
114947                               /organization/asys-2   
114948  /organization/novatiff-reklam-ve-tantm-hizmetl...   

                                           name  \
0                                         #fame   
2                                      :qounter   
3                        (the) one of them,inc.   
4                                       0-6.com   
6         